File: frontend/src/lib/screener.ts
import {
  calculateAtr,
  calculateBollingerBands,
  calculateMacd,
  calculateRsi,
  calculateTradeValue,
  calculateVolumeSpike,
  calculateVwap,
  simpleMovingAverage,
  type BollingerBand,
  type OhlcvBar,
  type VolumeSpike,
} from "./indicators.ts";

export type ScreenerStrategy = "BSJP" | "BPJS";

export type ScreenerInput = {
  current: OhlcvBar;
  previous: OhlcvBar;
  priceMa5: number;
  atr: number | null;
  rsi: number | null;
  macd: number | null;
  macdSignal: number | null;
  macdHistogram: number | null;
  bollinger: BollingerBand;
  vwap: number | null;
  volumeSpike: VolumeSpike;
};

export type ScreenerCriteria = {
  priceVsPrevious: boolean;
  priceVsMa5: boolean;
  priceVsOpen?: boolean;
  volumeVsPrevious: boolean;
  valueThreshold: boolean;
};

export type TradeLevels = {
  entry: number;
  stopLoss: number;
  takeProfit: number;
  exit: number;
  riskPerShare: number;
  rewardPerShare: number;
};

export type ScreenerCandidate = ScreenerInput & {
  strategy: ScreenerStrategy;
  criteria: ScreenerCriteria;
  score: number;
  value: number;
  levels: TradeLevels;
};

export type ScreenerResult = {
  bsjp: ScreenerCandidate[];
  bpjs: ScreenerCandidate[];
  all: ScreenerCandidate[];
};

export type ScreenerOptions = {
  limit?: number;
  minDailyValue?: number;
  atrMultiplier?: number;
  riskRewardRatio?: number;
};

const MIN_DAILY_VALUE = 5_000_000_000;
const DEFAULT_ATR_MULTIPLIER = 1;
const DEFAULT_RISK_REWARD_RATIO = 2;

export function buildScreenerInput(bars: OhlcvBar[]): ScreenerInput | null {
  if (bars.length < 6) {
    return null;
  }

  const current = bars.at(-1);
  const previous = bars.at(-2);

  if (!current || !previous) {
    return null;
  }

  const closes = bars.map((bar) => bar.close);
  const volumes = bars.map((bar) => bar.volume);
  const macd = calculateMacd(closes);
  const latestIndex = bars.length - 1;

  return {
    current,
    previous,
    priceMa5: simpleMovingAverage(closes, 5) ?? current.close,
    atr: calculateAtr(bars).at(-1) ?? null,
    rsi: calculateRsi(closes).at(-1) ?? null,
    macd: macd.macd.at(-1) ?? null,
    macdSignal: macd.signal.at(-1) ?? null,
    macdHistogram: macd.histogram.at(-1) ?? null,
    bollinger: calculateBollingerBands(closes).at(-1) ?? {
      middle: null,
      upper: null,
      lower: null,
    },
    vwap: calculateVwap(bars).at(-1) ?? null,
    volumeSpike: calculateVolumeSpike(volumes).at(latestIndex) ?? {
      ratio: null,
      isSpike: false,
    },
  };
}

export function isBsjpCandidate(
  input: ScreenerInput,
  minDailyValue = MIN_DAILY_VALUE,
): boolean {
  return Object.values(getBsjpCriteria(input, minDailyValue)).every(Boolean);
}

export function isBpjsCandidate(
  input: ScreenerInput,
  minDailyValue = MIN_DAILY_VALUE,
): boolean {
  return Object.values(getBpjsCriteria(input, minDailyValue)).every(Boolean);
}

export function screenStock(
  bars: OhlcvBar[],
  options: ScreenerOptions = {},
): ScreenerCandidate[] {
  const input = buildScreenerInput(bars);

  if (!input) {
    return [];
  }

  const minDailyValue = options.minDailyValue ?? MIN_DAILY_VALUE;
  const candidates: ScreenerCandidate[] = [];

  if (isBsjpCandidate(input, minDailyValue)) {
    candidates.push(createCandidate("BSJP", input, options));
  }

  if (isBpjsCandidate(input, minDailyValue)) {
    candidates.push(createCandidate("BPJS", input, options));
  }

  return candidates;
}

export function screenStocks(
  marketData: Record<string, OhlcvBar[]>,
  options: ScreenerOptions = {},
): ScreenerResult {
  const limit = options.limit ?? 5;
  const all = Object.values(marketData)
    .flatMap((bars) => screenStock(bars, options))
    .sort((a, b) => b.score - a.score);
  const bsjp = all
    .filter((candidate) => candidate.strategy === "BSJP")
    .slice(0, limit);
  const bpjs = all
    .filter((candidate) => candidate.strategy === "BPJS")
    .slice(0, limit);

  return {
    bsjp,
    bpjs,
    all: [...bsjp, ...bpjs].sort((a, b) => b.score - a.score),
  };
}

export function rankCandidates(
  inputs: ScreenerInput[],
  limit = 5,
): ScreenerCandidate[] {
  return inputs
    .flatMap((input) => {
      const candidates: ScreenerCandidate[] = [];

      if (isBsjpCandidate(input)) {
        candidates.push(createCandidate("BSJP", input));
      }

      if (isBpjsCandidate(input)) {
        candidates.push(createCandidate("BPJS", input));
      }

      return candidates;
    })
    .sort((a, b) => b.score - a.score)
    .slice(0, limit);
}

export function calculateTradeLevels(
  input: ScreenerInput,
  options: ScreenerOptions = {},
): TradeLevels {
  const entry = input.current.close;
  const atrMultiplier = options.atrMultiplier ?? DEFAULT_ATR_MULTIPLIER;
  const riskRewardRatio = options.riskRewardRatio ?? DEFAULT_RISK_REWARD_RATIO;
  const atrRisk = input.atr === null ? entry * 0.02 : input.atr * atrMultiplier;
  const riskPerShare = Math.max(atrRisk, entry * 0.005);
  const rewardPerShare = riskPerShare * riskRewardRatio;
  const stopLoss = entry - riskPerShare;
  const takeProfit = entry + rewardPerShare;

  return {
    entry,
    stopLoss,
    takeProfit,
    exit: takeProfit,
    riskPerShare,
    rewardPerShare,
  };
}

function createCandidate(
  strategy: ScreenerStrategy,
  input: ScreenerInput,
  options: ScreenerOptions = {},
): ScreenerCandidate {
  const minDailyValue = options.minDailyValue ?? MIN_DAILY_VALUE;
  const criteria =
    strategy === "BSJP"
      ? getBsjpCriteria(input, minDailyValue)
      : getBpjsCriteria(input, minDailyValue);

  return {
    ...input,
    strategy,
    criteria,
    value: calculateTradeValue(input.current.close, input.current.volume),
    score: calculateScore(input, strategy),
    levels: calculateTradeLevels(input, options),
  };
}

function getBsjpCriteria(
  input: ScreenerInput,
  minDailyValue: number,
): ScreenerCriteria {
  const { current, previous, priceMa5 } = input;
  const value = calculateTradeValue(current.close, current.volume);

  return {
    priceVsPrevious: current.close >= 1.05 * previous.close,
    priceVsMa5: current.close >= priceMa5,
    volumeVsPrevious: current.volume >= 1.2 * previous.volume,
    valueThreshold: value > minDailyValue,
  };
}

function getBpjsCriteria(
  input: ScreenerInput,
  minDailyValue: number,
): ScreenerCriteria {
  const { current, previous, priceMa5 } = input;
  const value = calculateTradeValue(current.close, current.volume);

  return {
    priceVsPrevious: current.close >= 1.05 * previous.close,
    priceVsMa5: current.close >= priceMa5,
    priceVsOpen: current.close >= current.open,
    volumeVsPrevious: current.volume >= 0.2 * previous.volume,
    valueThreshold: value > minDailyValue,
  };
}

function calculateScore(input: ScreenerInput, strategy: ScreenerStrategy): number {
  const priceMomentum = input.current.close / input.previous.close;
  const volumeMomentum = input.current.volume / Math.max(input.previous.volume, 1);
  const valueScore =
    calculateTradeValue(input.current.close, input.current.volume) / MIN_DAILY_VALUE;
  const rsiScore = input.rsi === null ? 0 : Math.min(input.rsi / 100, 1);
  const macdScore = input.macdHistogram === null ? 0 : Math.max(input.macdHistogram, 0);
  const vwapScore =
    input.vwap === null ? 0 : input.current.close / Math.max(input.vwap, 1);
  const spikeScore = input.volumeSpike.ratio ?? 0;
  const strategyWeight = strategy === "BSJP" ? 1.05 : 1;

  return (
    (priceMomentum * 2 +
      volumeMomentum +
      Math.min(valueScore, 5) +
      rsiScore +
      macdScore +
      vwapScore +
      spikeScore) *
    strategyWeight
  );
}

File: backend/app/core/screener.py
"""Logika screener IDX — port PERSIS dari frontend/src/lib/screener.ts.

Murni (tidak menyentuh DB / ML / cache) sehingga mudah diuji & dipakai ulang.
Dua strategi intraday:
  - BSJP (Beli Sore Jual Pagi)  — butuh lonjakan harga + volume kuat
  - BPJS (Beli Pagi Jual Sore)  — sama, plus close >= open, syarat volume longgar

Ambang & rumus disamakan baris-per-baris dengan TS (lihat komentar per blok).
Indikator dihitung ulang dari bar via app.core.indicators agar identik dengan
perhitungan client-side Phase 1.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, Protocol

from app.core import indicators

Strategy = Literal["BSJP", "BPJS"]

# Ambang default (mirror screener.ts).
MIN_DAILY_VALUE = 5_000_000_000  # Rp5 miliar turnover harian minimum
DEFAULT_ATR_MULTIPLIER = 1.0
DEFAULT_RISK_REWARD_RATIO = 2.0

# Screener butuh minimal 6 bar (current + previous + MA5 window).
MIN_BARS = 6


class OhlcvBar(Protocol):
    open: float
    high: float
    low: float
    close: float
    volume: float


@dataclass(frozen=True)
class ScreenerInput:
    current_open: float
    current_close: float
    current_volume: float
    previous_close: float
    previous_volume: float
    price_ma5: float
    atr: float | None
    rsi: float | None
    macd: float | None
    macd_signal: float | None
    macd_histogram: float | None
    bb_upper: float | None
    bb_middle: float | None
    bb_lower: float | None
    vwap: float | None
    volume_spike_ratio: float | None


@dataclass(frozen=True)
class TradeLevels:
    entry: float
    stop_loss: float
    take_profit: float
    exit: float
    risk_per_share: float
    reward_per_share: float


@dataclass(frozen=True)
class ScreenerCandidate:
    ticker: str
    strategy: Strategy
    criteria: dict[str, bool]
    score: float
    value: float
    levels: TradeLevels
    inp: ScreenerInput


def _trade_value(price: float, volume: float) -> float:
    return price * volume


def build_screener_input(bars: list[OhlcvBar]) -> ScreenerInput | None:
    """Rakit snapshot indikator bar terakhir. None bila data < MIN_BARS."""
    if len(bars) < MIN_BARS:
        return None

    current = bars[-1]
    previous = bars[-2]

    closes = [bar.close for bar in bars]
    volumes = [float(bar.volume) for bar in bars]
    macd = indicators.calculate_macd(closes)
    band = indicators.calculate_bollinger_bands(closes)[-1]
    spike = indicators.calculate_volume_spike(volumes)[-1]

    return ScreenerInput(
        current_open=current.open,
        current_close=current.close,
        current_volume=float(current.volume),
        previous_close=previous.close,
        previous_volume=float(previous.volume),
        price_ma5=indicators.simple_moving_average(closes, 5) or current.close,
        atr=indicators.calculate_atr(bars)[-1],
        rsi=indicators.calculate_rsi(closes)[-1],
        macd=macd.macd[-1],
        macd_signal=macd.signal[-1],
        macd_histogram=macd.histogram[-1],
        bb_upper=band.upper,
        bb_middle=band.middle,
        bb_lower=band.lower,
        vwap=indicators.calculate_vwap(bars)[-1],
        volume_spike_ratio=spike.ratio,
    )


def get_bsjp_criteria(inp: ScreenerInput, min_daily_value: float) -> dict[str, bool]:
    value = _trade_value(inp.current_close, inp.current_volume)
    return {
        "price_vs_previous": inp.current_close >= 1.05 * inp.previous_close,
        "price_vs_ma5": inp.current_close >= inp.price_ma5,
        "volume_vs_previous": inp.current_volume >= 1.2 * inp.previous_volume,
        "value_threshold": value > min_daily_value,
    }


def get_bpjs_criteria(inp: ScreenerInput, min_daily_value: float) -> dict[str, bool]:
    value = _trade_value(inp.current_close, inp.current_volume)
    return {
        "price_vs_previous": inp.current_close >= 1.05 * inp.previous_close,
        "price_vs_ma5": inp.current_close >= inp.price_ma5,
        "price_vs_open": inp.current_close >= inp.current_open,
        "volume_vs_previous": inp.current_volume >= 0.2 * inp.previous_volume,
        "value_threshold": value > min_daily_value,
    }


def _criteria_for(
    strategy: Strategy, inp: ScreenerInput, min_daily_value: float
) -> dict[str, bool]:
    return (
        get_bsjp_criteria(inp, min_daily_value)
        if strategy == "BSJP"
        else get_bpjs_criteria(inp, min_daily_value)
    )


def calculate_score(inp: ScreenerInput, strategy: Strategy) -> float:
    """Skor komposit — port dari calculateScore (screener.ts)."""
    price_momentum = inp.current_close / inp.previous_close
    volume_momentum = inp.current_volume / max(inp.previous_volume, 1)
    value_score = _trade_value(inp.current_close, inp.current_volume) / MIN_DAILY_VALUE
    rsi_score = 0.0 if inp.rsi is None else min(inp.rsi / 100, 1)
    macd_score = 0.0 if inp.macd_histogram is None else max(inp.macd_histogram, 0)
    vwap_score = 0.0 if inp.vwap is None else inp.current_close / max(inp.vwap, 1)
    spike_score = inp.volume_spike_ratio or 0.0
    strategy_weight = 1.05 if strategy == "BSJP" else 1.0

    return (
        price_momentum * 2
        + volume_momentum
        + min(value_score, 5)
        + rsi_score
        + macd_score
        + vwap_score
        + spike_score
    ) * strategy_weight


def calculate_trade_levels(
    inp: ScreenerInput,
    atr_multiplier: float = DEFAULT_ATR_MULTIPLIER,
    risk_reward_ratio: float = DEFAULT_RISK_REWARD_RATIO,
) -> TradeLevels:
    entry = inp.current_close
    atr_risk = entry * 0.02 if inp.atr is None else inp.atr * atr_multiplier
    risk_per_share = max(atr_risk, entry * 0.005)
    reward_per_share = risk_per_share * risk_reward_ratio
    return TradeLevels(
        entry=entry,
        stop_loss=entry - risk_per_share,
        take_profit=entry + reward_per_share,
        exit=entry + reward_per_share,
        risk_per_share=risk_per_share,
        reward_per_share=reward_per_share,
    )


def _create_candidate(
    ticker: str,
    strategy: Strategy,
    inp: ScreenerInput,
    min_daily_value: float,
    atr_multiplier: float,
    risk_reward_ratio: float,
) -> ScreenerCandidate:
    return ScreenerCandidate(
        ticker=ticker,
        strategy=strategy,
        criteria=_criteria_for(strategy, inp, min_daily_value),
        value=_trade_value(inp.current_close, inp.current_volume),
        score=calculate_score(inp, strategy),
        levels=calculate_trade_levels(inp, atr_multiplier, risk_reward_ratio),
        inp=inp,
    )


def screen_bars(
    ticker: str,
    bars: list[OhlcvBar],
    min_daily_value: float = MIN_DAILY_VALUE,
    atr_multiplier: float = DEFAULT_ATR_MULTIPLIER,
    risk_reward_ratio: float = DEFAULT_RISK_REWARD_RATIO,
) -> list[ScreenerCandidate]:
    """Kembalikan kandidat (0..2) untuk satu ticker, satu per strategi yang lolos."""
    inp = build_screener_input(bars)
    if inp is None:
        return []

    candidates: list[ScreenerCandidate] = []
    for strategy in ("BSJP", "BPJS"):
        criteria = _criteria_for(strategy, inp, min_daily_value)  # type: ignore[arg-type]
        if all(criteria.values()):
            candidates.append(
                _create_candidate(
                    ticker,
                    strategy,  # type: ignore[arg-type]
                    inp,
                    min_daily_value,
                    atr_multiplier,
                    risk_reward_ratio,
                )
            )
    return candidates
